Add post on re-engineering CI/CD pipelines for SciPy

HarshCasper opened the pull request:
<!--Hi there! If you aren't submitting a blog post, you can delete this message.

Thank you for submitting a blog post! We want to make sure content produced for Quansight Labs is accessible, so please address the the following guidelines in your post.-->

## Text styling

- [X] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [X] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [X] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [X] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [X] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [X] If there are emojis, there are not more than three in a row.
- [X] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [X] If it were to be read as plain text, the blog still makes sense and no information is missing.

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,160 @@
+
+GitHub Actions features a vast number of architectures, operating systems, runners, and more. With the aid of GitHub Actions, we can directly use the GitHub API from our workflows themselves. GitHub Actions also features a [vibrant marketplace](https://github.com/marketplace?type=actions) with applications centered around making the usage of GitHub Actions more straightforward for developers.
+
+We can also emulate local CI runs using a tool like [Act](https://github.com/nektos/act) to validate our GitHub Actions workflow. It allows us to run the entire workflow (including build and test) on our local machine, using a Docker image without initially testing it on the CI. Apart from this, whenever fork a repository, the actions automatically get forked. It allows us to test on our fork and ...
```
```suggestion
We can also emulate local CI runs using a tool like [Act](https://github.com/nektos/act) to validate our GitHub Actions workflow. It allows us to run the entire workflow (including build and test) on our local machine, using a Docker image without initially testing it on the CI. Apart from this, whenever we fork a repository, the actions automatically get forked. It allows us to test on our fork and makes it relatively straightforward to use.
```

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,160 @@
+
+## Integrating build caching using Ccache
+
+Today, it takes around ~35 minutes for Linux and macOS tests to pass on GitHub Actions. With Scipy’s migration efforts towards the Meson build system, it takes around ~25 minutes for the CI to pass the build, install, and tests. Our primary focus was to reduce the overall time to achieve this. But finding workarounds for the same was particularly challenging.
```
Just to be explicit:
```suggestion
Today, it takes around ~35 minutes for SciPy's Linux and macOS tests to pass on GitHub Actions. With Scipy’s migration efforts towards the Meson build system, it takes around ~25 minutes for the CI to pass the build, install, and run tests. Our primary focus was to reduce the overall time to achieve this. But finding workarounds for the same was particularly challenging.
```

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,160 @@
+
+With the GitHub Action workflows available for Linux, it seemed reasonable that we would like to test the same on macOS. This [issue](https://github.com/rgommers/scipy/issues/64) by [Rohit Goswami](https://github.com/HaoZeke), where he documents the problems with the macOS build, made sense that we can implement a workflow with little to almost no tweaks. However, I wanted to implement the conda- ...
+
+GitHub Actions allows us to perform macOS builds quite generously, and creating a workflow for the same was relatively easy. After the same, we had to make a couple of iterations to verify if the compiler caching exactly works and how performant it is compared to a fresh build. I also implemented conda caching for faster installs using [Mamba](https://github.com/mamba-org/mamba), which reduced th ...
```
Maybe just to avoid repetition:
```suggestion
GitHub Actions allows us to perform macOS builds quite generously, and creating a workflow for the same was relatively easy. After this, we had to make a couple of iterations to verify if the compiler caching exactly works and how performant it is compared to a fresh build. I also implemented conda caching for faster installs using [Mamba](https://github.com/mamba-org/mamba), which reduced the effective build time by 30%.
```

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,160 @@
+
+- Having dedicated hardware for the same can lead to abuse and unnecessary noise.
+
+We decided to use either [Cachegrind](https://valgrind.org/docs/manual/cg-manual.html) or [Jaime’s approach](https://labs.quansight.org/blog/2021/08/github-actions-benchmarks/) to tackle all of these problems, which talks about relative benchmarking on GitHub Actions. Both Scikit Image and SciPy use Airspeed Velocity (`asv`) for benchmarking. This rules out the use of cachegrind since `asv` is no ...
```
```suggestion
We decided to use either [Cachegrind](https://valgrind.org/docs/manual/cg-manual.html) or [Jaime’s approach](https://labs.quansight.org/blog/2021/08/github-actions-benchmarks/) to tackle all of these problems, which talks about relative benchmarking on GitHub Actions. Both scikit-image and SciPy use Airspeed Velocity (`asv`) for benchmarking. This rules out the use of cachegrind since `asv` is not compatible with the same.
```

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,160 @@
+
+- macOS (ARM64): Requires more effort with no options available on GitHub Actions.
+
+We also implemented a sampling strategy to match the versions of Python and NumPy. In addition, the special purpose jobs included are for Linting, docs build, Gitpod build, Benchmarking, Pythran, MyPy, and more.
```
```suggestion
We also implemented a sampling strategy to match the versions of Python and NumPy. In addition, the special purpose jobs included are for linting, docs build, Gitpod build, Benchmarking, Pythran, MyPy, and more.
```

melissawm reviewed (commented):
Hello, @HarshCasper ! I left a couple of suggestions mostly for readability. One high-level comment is that this post is very technical and detailed. I think that is ok for the audience we expect at the Labs blog, but it might be worth adding a note in the beginning that this will focus on technical aspects and might required some previous familiarity with CI pipelines and infrastructure.

HarshCasper commented:
Hi @melissawm 

Thanks a lot for the review and the comments!

I have fixed all of them here and added some pictures with associated alt tags. Please let me know if more changes are required on the same. Thanks for the feedback :smile:

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,211 @@
+
+[SciPy](https://github.com/scipy/scipy) is quite an old and mature project, which is being used across the PyData community. The project currently uses a variety of continuous integration services and the Setuptools build system. As the project is growing and `distutils` will be deprecated as part of [PEP 632](https://www.python.org/dev/peps/pep-0632/), it was reasonable to migrate to a new build ...
+
+Scipy has been previously using [TravisCI](https://travis-ci.org/), [CircleCI](https://circleci.com/), and [AppVeyor](https://www.appveyor.com/) across its continuous integration pipelines. Recently, [GitHub Actions](https://github.com/features/actions), the native CI/CD service offered by GitHub, has been picked up by many projects. My project was centered around building and re-engineering the  ...
```
nit:
```suggestion
SciPy has been previously using [TravisCI](https://travis-ci.org/), [CircleCI](https://circleci.com/), and [AppVeyor](https://www.appveyor.com/) across its continuous integration pipelines. Recently, [GitHub Actions](https://github.com/features/actions), the native CI/CD service offered by GitHub, has been picked up by many projects. My project was centered around building and re-engineering the continuous integration pipelines for SciPy using GitHub Actions to be made compatible with the new build system.
```

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,211 @@
+
+<p align="center">
+    <img
+     alt="The project will enable the developers to test out the all-new Meson build system on the GitHub Actions CI and lays down the further scope for developing an entire CI matrix for build, test and release for SciPy."
```
The alt-text should include a description of the image. In this case, it could be something like "The Quansight Labs and SciPy logos side by side." There is probably already a description of those logos somewhere, we could even link to that I guess.

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,211 @@
+
+<p align="center">
+    <img
+     alt="Meme depicting the state of CI service providers today. While many projects
```
This also needs to be more descriptive. For example: "A meme showing two panels. In the top one, a man says "Life is good" while pointing at the camera. In the bottom panel, the same man smiles and says "But it can be better". Next to the meme, a series of logos for CI services are shown, including the logos for TravisCI, CircleCI and github actions" (I don't actually know all the logos!)

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,211 @@
+
+<p align="center">
+    <img
+     alt="A snapshot from the Visual Studio Code demonstrating the usage of `act` locally to test out the GitHub Action workflows. In the picture, we are running the `test_meson` workflow which runs the CI job locally to build, install and test SciPy locally."
```
This one is good!

melissawm commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,211 @@
+
+<p align="center">
+    <img
+     alt="Meson build system natively supports Ccache for compiler target caching. One of the project milestones required us to reduce the overall CI build time. With the help of compiler caching and the Cache action provided by GitHub Actions we were able to significantly tackle the total CI build time."
```
Here, I think we need to emphasize the information you want to convey with the images: the difference in build time.

melissawm reviewed (commented):
Overall looks good! I left observations about the alt-text for the images. The alt-text should be descriptive of the image itself, and should not contain extra information. Some of the alt-text you wrote is quite good and could even be incorporated into the text, but it needs to be reformulated into actual alt-text for the images. [Here's a nice guide about writing alt-text](https://webaim.org/techniques/alttext/).

melissawm reviewed (approved):
This is looking good, thanks @HarshCasper ! I have no other comments :)

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+[SciPy](https://github.com/scipy/scipy) is quite an old and mature project, which is being used across the PyData community. The project currently uses a variety of continuous integration services and the Setuptools build system. As the project is growing and `distutils` will be deprecated as part of [PEP 632](https://www.python.org/dev/peps/pep-0632/), it was reasonable to migrate to a new build ...
+
+SciPy has been previously using [TravisCI](https://travis-ci.org/), [CircleCI](https://circleci.com/), and [AppVeyor](https://www.appveyor.com/) across its continuous integration pipelines. Recently, [GitHub Actions](https://github.com/features/actions), the native CI/CD service offered by GitHub, has been picked up by many projects. My project was centered around building and re-engineering the  ...
```
It would be good to point out that SciPy was already using GHA in its main repo - here we're just increasing the use of it. In https://github.com/MacPython/scipy-wheels we're not using GHA at all yet, so there's more room to improve.

HarshCasper replied:
Refactored the paragraph and added a note that it was about increasing the adoption of the GHA usage.

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+- SciPy currently features benchmarks that are not automated on the CI. We want to automate this process on the CI itself by tagging Pull Requests with specific labels or triggering the same during releases.
+
+- Scipy currently supports X86, X64, PPC64, and ARM64 architecture through the [Azure DevOps](https://azure.microsoft.com/en-in/services/devops/) pipelines. We want to ensure that the compatibility remains the same with the new build system.
```
I'd spell the first two as `x86` and `x86_64`.

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+`act` is a tool offered by Nektos which provides a handy way to run GitHub Actions locally. It gives a quick way to validate your changes on the CI locally, without committing/pushing your changes to the workflows to trigger and validate them. It leads to fast feedback and compatibility as a local task runner to validate all our CI jobs.
+
+You can try out `act` yourself by cloning [Ralf’s fork](https://github.com/rgommers/scipy) and referencing the official documentation for the [`act` usage](https://github.com/rgommers/scipy/blob/meson/ACT-USAGE.md). It keeps the GitHub Action notifications clutter-free and reduces the chance of accidentally triggering any failure during debugging.
```
I'd remove "official" here. When you say "official documentation" I'd think that's the upstream docs of the tool itself, not our docs in the scipy repo.

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+- `__config__.py`: This file is generated by NumPy and should use a version of NumPy at build time that is compatible with all potential runtime versions of NumPy.
+
+- `version.py`: This file is dependent on the Git tags. We were looking to instantiate a file in the build directory that represents the current one.
```
"the Git tags" -> "Git commit hash"

rgommers replied:
"We were looking to instantiate a file in the build directory that represents the current one." ->

"We needed to generate a file in the build directory instead of in-place, because Meson does not allow writing output into the source tree that is being built."

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+While digging around the documentation, I discovered that `ccache` is natively supported by Meson and hence would be a better choice. I found that the build targets are exactly being cached by debugging around, and I can use the cache action to cache them. Finally, I made it work, thus effectively reducing the build time by nearly ~70% and reducing the CI time.
+
+In addition, I also worked on caching PyPi dependencies using the cache actions that allowed us to shave off some of the time required during the Python dependency installation. [PR #63](https://github.com/rgommers/scipy/pull/63) containing the commits for the code necessary to set up build caching as part of the GitHub Action workflow for Linux.
```
"PyPi" - "PyPI"

HarshCasper replied:
Fixed.

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+- macOS (ARM64): Requires more effort with no options available on GitHub Actions.
+
+We also implemented a sampling strategy to match the versions of Python and NumPy. In addition, the special purpose jobs included are for linting, docs build, Gitpod build, Benchmarking, Pythran, MyPy, and more.
```
minor: `Benchmarking` should probably start with a small `b`

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+## What did I learn from this project?
+
+The summer internship at Quansight Labs has been the busiest time of my life for all good reasons. I learned a great deal about GitHub Actions, build systems, and PyData projects through the same. I also learned about testing and benchmarking, along with Docker by a significant margin. I had the basic idea for some of them and have worked on them in hobby projects, but implementing them myself on ...
```
the "by a significant margin" seems a little off to me. Maybe "along with a more in-depth understanding of Docker" or something like that?

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+I want to thank my mentor, John Lee, for his impeccable mentorship and feedback. Without his help and support, all this would not have been possible. Above all, his skill in drawing from past experiences to future model outcomes has been helpful for me to understand things beyond open source and software development itself.
+
+I would also like to thank Ralf Gommers, who has been instrumental in my project work. His feedback, continued support, and long-term vision contributed significantly to my project. [Anirudh Dagar](https://github.com/AnirudhDagar/), my co-intern, helped me understand more about SciPy and helping with my first PR reviews. Jaime helped me considerably understand benchmarking, and getting to know mo ...
```
You're using the full name for everyone here except Jaime - I suggest adding his last name too.

HarshCasper replied:
I have used his full name before so I have reverted to use the first name for everyone in the `Acknowledgements` section.

rgommers commented on posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md:
```diff
@@ -0,0 +1,209 @@
+
+[PR #67](https://github.com/rgommers/scipy/pull/67) contains the commits for the code required to set up the GitHub Action workflow for automated CI benchmarking using `asv` and artifact upload.
+
+## Strategizing the CI Matrix for SciPy
```
You could consider using a screenshot of the CI matrix, if you remove the yellow lines and make it a bit more compact, this could look nice/useful:

<img width="781" alt="image" src="https://user-images.githubusercontent.com/98330/136621128-425a6093-2f77-4c78-b580-601c60dde5cb.png">

rgommers reviewed (commented):
This looks great @HarshCasper! A number of comments from me still - most are small though and should not be hard to address. This is about ready to go, it reads well.
